File: puppeteer/ecom.js
const puppeteer = require('puppeteer');
require('dotenv').config(); // Load environment variables from a .env file

puppeteer.launch({ headless: false, slowMo: 50 }) // Launch Puppeteer in non-headless mode with a slight slowdown for visibility
  .then(browser => browser.newPage() // Create a new page in the opened browser instance
    .then(page => {
      return page.goto('https://www.saucedemo.com/') // Navigate to the URL
        .then(() => page.$eval('.login_logo', el => el.textContent)) // Extract text content of the .login_logo element
        .then(logoText => {
          // Validate if the logo text includes 'Swag Labs'
          if (logoText.includes('Swag Labs')) {
            console.log('Text validation passed.');
          } else {
            console.log('Text validation failed.');
          }
          // Wait for the username input field to appear
          return page.waitForSelector('#user-name');
        })
        .then(() => page.type('#user-name', process.env.USERNAME)) // Type the username fetched from environment variables
        .then(() => page.waitForSelector('#password')) // Wait for the password input field to appear
        .then(() => page.type('#password', process.env.PASSWORD)) // Type the password fetched from environment variables
        .then(() => page.waitForSelector('#login-button')) // Wait for the login button to appear
        .then(() => page.click('#login-button')) // Click on the login button
        .then(() => {
          console.log('Logged in successfully.');
          // Wait for the title element to appear on the inventory page
          return page.waitForSelector('.title');
        })
        .then(() => page.$eval('.title', el => el.textContent)) // Extract text content of the title element
        .then(titleText => {
          // Validate if the title text includes 'Products'
          if (titleText.includes('Products')) {
            console.log('Products text validation passed.');
          } else {
            console.log('Products text validation failed.');
          }
          // Wait for the product name elements to appear
          return page.waitForSelector('.inventory_item_name');
        })
        .then(() => page.$$eval('.inventory_item_name', elements => elements.map(el => el.textContent))) // Extract text content of all product name elements
        .then(productNames => {
          const productIndex = productNames.indexOf('Sauce Labs Onesie');
          if (productIndex !== -1) {
            // Calculate the selector for the product based on its index
            const productSelector = `.inventory_item:nth-child(${productIndex + 1}) .inventory_item_name`;
            // Click on the product based on the generated selector
            return page.click(productSelector)
              .then(() => {
                console.log('Clicked on "Sauce Labs Onesie".');
                // Wait for the product details name element to appear
                return page.waitForSelector('.inventory_details_name');
              })
              .then(() => page.$eval('.inventory_details_name', el => el.textContent)) // Extract text content of the product details name element
              .then(productName => {
                // Validate if the product details name includes 'Sauce Labs Onesie'
                if (productName.includes('Sauce Labs Onesie')) {
                  console.log('Product name validation passed.');
                } else {
                  console.log('Product name validation failed.');
                }
                // Extract text content of the product details price element
                return page.$eval('.inventory_details_price', el => el.textContent);
              })
              .then(priceText => {
                // Validate if the price text includes '$'
                if (priceText.includes('$')) {
                  console.log('Price validation passed.');
                } else {
                  console.log('Price validation failed.');
                }
                // Wait for the 'Add to cart' button to appear
                return page.waitForSelector('#add-to-cart');
              })
              .then(() => page.click('#add-to-cart')) // Click on the 'Add to cart' button
              .then(() => {
                console.log('Clicked on "Add to cart" button.');
                // Wait for the shopping cart link to appear
                return page.waitForSelector('a.shopping_cart_link');
              })
              .then(() => page.click('a.shopping_cart_link')) // Click on the shopping cart link
              .then(() => {
                console.log('Clicked on shopping cart icon.');
                // Wait for the cart item elements to appear
                return page.waitForSelector('.cart_item');
              })
              .then(() => page.$$eval('.cart_item', elements => elements.map(el => el.textContent))) // Extract text content of all cart item elements
              .then(cartItems => {
                // Validate if 'Sauce Labs Onesie' is found in the cart items
                if (cartItems.some(item => item.includes('Sauce Labs Onesie'))) {
                  console.log('Product "Sauce Labs Onesie" found in the cart.');
                } else {
                  console.log('Product "Sauce Labs Onesie" not found in the cart.');
                }
                // Wait for the 'Checkout' button to appear
                return page.waitForSelector('#checkout');
              })
              .then(() => page.click('#checkout')) // Click on the 'Checkout' button
              .then(() => {
                console.log('Clicked on "Checkout" button.');
                // Wait for the title element to appear again
                return page.waitForSelector('.title');
              })
              .then(() => page.$eval('.title', el => el.textContent)) // Extract text content of the title element again
              .then(checkoutHeaderText => {
                // Validate if the checkout header text includes 'Checkout: Your Information'
                if (checkoutHeaderText.includes('Checkout: Your Information')) {
                  console.log('Checkout header text validation passed.');
                } else {
                  console.log('Checkout header text validation failed.');
                }
                // Wait for the 'First Name' input field to appear
                return page.waitForSelector('#first-name');
              })
              .then(() => page.type('#first-name', 'John')) // Type 'John' into the 'First Name' input field
              .then(() => page.waitForSelector('#last-name')) // Wait for the 'Last Name' input field to appear
              .then(() => page.type('#last-name', 'Doe')) // Type 'Doe' into the 'Last Name' input field
              .then(() => page.waitForSelector('#postal-code')) // Wait for the 'Postal Code' input field to appear
              .then(() => page.type('#postal-code', '12345')) // Type '12345' into the 'Postal Code' input field
              .then(() => page.waitForSelector('#continue')) // Wait for the 'Continue' button to appear
              .then(() => page.click('#continue')) // Click on the 'Continue' button
              .then(() => {
                console.log('Clicked on "Continue" button.');
                // Wait for the 'Finish' button to appear
                return page.waitForSelector('#finish');
              })
              .then(() => page.$eval('#finish', el => el.textContent)) // Extract text content of the 'Finish' button
              .then(finishButtonText => {
                // Validate if the 'Finish' button text includes 'Finish'
                if (finishButtonText.includes('Finish')) {
                  console.log('Finish button text validation passed.');
                } else {
                  console.log('Finish button text validation failed.');
                }
                // Wait for the 'Finish' button to appear again
                return page.waitForSelector('#finish');
              })
              .then(() => page.click('#finish')) // Click on the 'Finish' button
              .then(() => {
                console.log('Clicked on "Finish" button.');
                // Wait for the 'Complete' header to appear
                return page.waitForSelector('.complete-header');
              })
              .then(() => page.$eval('.complete-header', el => el.textContent)) // Extract text content of the 'Complete' header
              .then(thankYouText => {
                // Validate if the 'Thank you' text includes 'Thank you for your order!'
                if (thankYouText.includes('Thank you for your order!')) {
                  console.log('Thank you text validation passed.');
                } else {
                  console.log('Thank you text validation failed.');
                }
                // Wait for the 'Back to Products' button to appear
                return page.waitForSelector('#back-to-products');
              })
              .then(() => page.click('#back-to-products')) // Click on the 'Back to Products' button
              .then(() => {
                console.log('Clicked on "Back Home" button.');
                // Wait for the title element to appear again
                return page.waitForSelector('.title');
              })
              .then(() => page.$eval('.title', el => el.textContent)) // Extract text content of the title element again
              .then(productsHeaderText => {
                // Validate if the title text includes 'Products'
                if (productsHeaderText.includes('Products')) {
                  console.log('Products text validation passed.');
                } else {
                  console.log('Products text validation failed.');
                }

                // Additional steps for logging out
                // Wait for the burger menu button to appear
                return page.waitForSelector('.bm-burger-button', { visible: true });
              })
              .then(() => page.click('.bm-burger-button')) // Click on the burger menu button
              .then(() => {
                console.log('Clicked on lines icon');
                // Wait for the logout link to appear
                return page.waitForSelector('a#logout_sidebar_link.bm-item.menu-item', { visible: true });
              })
              .then(() => page.evaluate(() => {
                // Scroll the logout link into view
                const logoutLink = document.querySelector('a#logout_sidebar_link.bm-item.menu-item');
                logoutLink.scrollIntoView();
              }))
              .then(() => page.$eval('a#logout_sidebar_link.bm-item.menu-item', (elem) => {
                // Evaluate if the logout link is clickable
                const rect = elem.getBoundingClientRect();
                return (
                  rect.width > 0 &&
                  rect.height > 0 &&
                  window.getComputedStyle(elem).visibility !== 'hidden' &&
                  window.getComputedStyle(elem).display !== 'none'
                );
              }))
              .then(isClickable => {
                if (isClickable) {
                  // Click on the logout link if it's clickable
                  return page.click('a#logout_sidebar_link.bm-item.menu-item').then(() => {
                    console.log('Clicked on the Logout link.');
                    // Wait for the login button to appear again
                    return page.waitForSelector('#login-button');
                  });
                } else {
                  throw new Error('Logout link is not clickable');
                }
              })
              .then(() => page.$eval('#login-button', el => el.value)) // Extract the text content of the login button
              .then(loginButtonValue => {
                // Validate if the login button text is 'Login'
                if (loginButtonValue === 'Login') {
                  console.log('The login button text is correct.');
                } else {
                  console.log('The login button text is incorrect. Found:', loginButtonValue);
                }
                console.log('Script execution completed.');
                // Close the browser instance
                return browser.close();
              })
              .catch(err => {
                // Handle and log errors during script execution
                console.error('Error occurred:', err);
                // Close the browser instance in case of error
                return browser.close();
              });
          } else {
            console.log('"Sauce Labs Onesie" not found.');
            // Close the browser instance if 'Sauce Labs Onesie' is not found
            return browser.close();
          }
        });
    }))
  .catch(err => {
    // Handle and log errors during Puppeteer launch or page creation
    console.error('Error occurred:', err);
  });

File: selenium/ecom.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Setup WebDriver
options = webdriver.ChromeOptions()
options.headless = False
options.add_argument('--start-maximized')
driver = webdriver.Chrome(options=options)

try:
    # Navigate to the URL
    driver.get('https://www.saucedemo.com/')

    # Validate if the logo text includes 'Swag Labs'
    logo_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'login_logo'))
    ).text
    if 'Swag Labs' in logo_text:
        print('Text validation passed.')
    else:
        print('Text validation failed.')

    # Enter username and password
    username = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, 'user-name'))
    )
    password = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, 'password'))
    )

    username_value = os.getenv('USERNAME')
    password_value = os.getenv('PASSWORD')
    
    if username_value is None or password_value is None:
        raise ValueError('Environment variables USERNAME and PASSWORD must be set.')

    username.send_keys(username_value)
    password.send_keys(password_value)

    # Click login button
    login_button = driver.find_element(By.ID, 'login-button')
    login_button.click()

    print('Logged in successfully.')

    # Validate if the title text includes 'Products'
    title_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'title'))
    ).text
    if 'Products' in title_text:
        print('Products text validation passed.')
    else:
        print('Products text validation failed.')

    # Click on 'Sauce Labs Onesie' product
    product_names = driver.find_elements(By.CLASS_NAME, 'inventory_item_name')
    if product_names is None:
        raise ValueError('No products found.')

    for product in product_names:
        if 'Sauce Labs Onesie' in product.text:
            product.click()
            break

    print('Clicked on "Sauce Labs Onesie".')

    # Validate if the product details name includes 'Sauce Labs Onesie'
    product_name = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'inventory_details_name'))
    ).text
    if 'Sauce Labs Onesie' in product_name:
        print('Product name validation passed.')
    else:
        print('Product name validation failed.')

    # Validate if the price text includes '$'
    price_text = driver.find_element(By.CLASS_NAME, 'inventory_details_price').text
    if '$' in price_text:
        print('Price validation passed.')
    else:
        print('Price validation failed.')

    # Click on 'Add to cart' button
    add_to_cart = driver.find_element(By.XPATH, '//*[@id="add-to-cart"]')
    add_to_cart.click()
    print('Clicked on "Add to cart" button.')

    # Click on the shopping cart link
    cart_link = driver.find_element(By.CLASS_NAME, 'shopping_cart_link')
    cart_link.click()
    print('Clicked on shopping cart icon.')

    # Validate if 'Sauce Labs Onesie' is found in the cart items
    cart_items = driver.find_elements(By.CLASS_NAME, 'cart_item')
    if cart_items is None:
        raise ValueError('No cart items found.')

    if any('Sauce Labs Onesie' in item.text for item in cart_items):
        print('Product "Sauce Labs Onesie" found in the cart.')
    else:
        print('Product "Sauce Labs Onesie" not found in the cart.')

    # Click on the 'Checkout' button
    checkout = driver.find_element(By.ID, 'checkout')
    checkout.click()
    print('Clicked on "Checkout" button.')

    # Validate if the checkout header text includes 'Checkout: Your Information'
    checkout_header_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'title'))
    ).text
    if 'Checkout: Your Information' in checkout_header_text:
        print('Checkout header text validation passed.')
    else:
        print('Checkout header text validation failed.')

    # Fill in checkout information
    driver.find_element(By.ID, 'first-name').send_keys('John')
    driver.find_element(By.ID, 'last-name').send_keys('Doe')
    driver.find_element(By.ID, 'postal-code').send_keys('12345')

    # Click on the 'Continue' button
    continue_button = driver.find_element(By.ID, 'continue')
    continue_button.click()
    print('Clicked on "Continue" button.')

    # Validate if the 'Finish' button text includes 'Finish'
    finish_button_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, 'finish'))
    ).text
    if 'Finish' in finish_button_text:
        print('Finish button text validation passed.')
    else:
        print('Finish button text validation failed.')

    # Click on the 'Finish' button
    finish_button = driver.find_element(By.ID, 'finish')
    finish_button.click()
    print('Clicked on "Finish" button.')

    # Validate if the 'Thank you' text includes 'Thank you for your order!'
    thank_you_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'complete-header'))
    ).text
    if 'Thank you for your order!' in thank_you_text:
        print('Thank you text validation passed.')
    else:
        print('Thank you text validation failed.')

    # Click on the 'Back Home' button
    back_home = driver.find_element(By.ID, 'back-to-products')
    back_home.click()
    print('Clicked on "Back Home" button.')

    # Validate if the title text includes 'Products'
    products_header_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'title'))
    ).text
    if 'Products' in products_header_text:
        print('Products text validation passed.')
    else:
        print('Products text validation failed.')

    # Additional steps for logging out
    burger_menu_button = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.CLASS_NAME, 'bm-burger-button'))
    )
    burger_menu_button.click()
    print('Clicked on lines icon')


    logout_link = WebDriverWait(driver, 10).until(
        EC.element_to_be_clickable((By.ID, 'logout_sidebar_link'))
    )
    logout_link.click()
    print('Clicked on the Logout link.')

    # Validate if the login button text is 'Login'
    login_button_text = WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, 'login-button'))
    ).get_attribute('value')
    if login_button_text == 'Login':
        print('The login button text is correct.')
    else:
        print('The login button text is incorrect. Found:', login_button_text)

    print('Script execution completed.')

except Exception as e:
    print(f'Error occurred: {e}')

finally:
    driver.quit()
